builtin/k8s/apply/platform.go

```diff
@@ -0,0 +1,302 @@
+	return p.Deploy
+}
+
+// DestroyFunc implements component.Destroyer
+func (p *Platform) DestroyFunc() interface{} {
+	return p.Destroy
+}
+
+// GenerationFunc implements component.Generation
+func (p *Platform) GenerationFunc() interface{} {
+	return p.Generation
+}
+
+// Deploy deploys to Kubernetes
+func (p *Platform) Deploy(
+	ctx context.Context,
+	log hclog.Logger,
+	deployConfig *component.DeploymentConfig,
+	ui terminal.UI,
+) (*Deployment, error) {
+	sg := ui.StepGroup()
+	defer sg.Wait()
+	s := sg.Add("")
+	defer func() { s.Abort() }()
+
+	deployment := &Deployment{
+		PruneLabel: p.config.PruneLabel,
+	}
+
+	s.Update("Executing kubectl apply...")
+	cmd, err := p.cmd(ctx, s, "apply", "-f", p.config.Path, "--prune", "-l", deployment.PruneLabel)
+	if err != nil {
+		return nil, err
+	}
+	if err := cmd.Run(); err != nil {
+		return nil, err
+	}
+	s.Done()
+
+	return deployment, nil
+}
+
+// Destroy
+func (p *Platform) Destroy(
+	ctx context.Context,
+	log hclog.Logger,
+	deployment *Deployment,
+	ui terminal.UI,
+) error {
+	sg := ui.StepGroup()
+	defer sg.Wait()
+	s := sg.Add("Executing kubectl to destroy...")
+	defer func() { s.Abort() }()
+
+	if deployment.PruneLabel == "" {
+		s.Update("No prune label on deployment. Not destroying.")
+		s.Done()
+		return nil
+	}
+
+	// Apply it
+	cmd, err := p.cmd(ctx, s, "delete", "all", "-l", deployment.PruneLabel)
+	if err != nil {
+		return err
+	}
+	if err := cmd.Run(); err != nil {
+		return err
+	}
+
+	s.Done()
+	return nil
+}
+
+// Generation returns the generation ID.
+func (p *Platform) Generation(
+	ctx context.Context,
+) ([]byte, error) {
+	// Static generation since we will always use the `prune_label` to
+	// automatically delete unused resources.
+	return []byte("kubernetes-apply"), nil
+}
+
+func (p *Platform) cmd(
+	ctx context.Context,
+	step terminal.Step,
+	subcmd string, args ...string,
+) (*exec.Cmd, error) {
+	path, err := exec.LookPath("kubectl")
+	if err != nil {
+		return nil, status.Errorf(codes.FailedPrecondition,
+			"kubectl could not be found on the PATH")
+	}
+
+	// Build the args we'll send to kubectl, which are:
+	// <subcmd> <flags> <flags + args provided to this func>
+	realArgs := []string{subcmd}
+	if p.config.KubeconfigPath != "" {
+		realArgs = append(realArgs, "-kubeconfig="+p.config.KubeconfigPath)
+	}
+	realArgs = append(realArgs, args...)
+
+	// Build our command
+	cmd := exec.CommandContext(ctx, path, realArgs...)
+
+	// Ensure output goes to our step
+	cmd.Stdout = step.TermOutput()
+	cmd.Stderr = cmd.Stdout
+	return cmd, nil
+}
+
+// Config is the configuration structure for the Platform.
+type Config struct {
+	// The path to the job specification to load.
+	Path string `hcl:"path,attr"`
+
+	// Prune label is the label to use to destroy resources that don't match.
+	PruneLabel string `hcl:"prune_label,attr"`
+
+	// KubeconfigPath is the path to the kubeconfig file.
+	KubeconfigPath string `hcl:"kubeconfig,optional"`
```

Comment:
Should we also support specifying a kube context here so that people won't have to make sure their env is right before running waypoint up?

Reply:
I like that idea! If we support it, folks can specify the exact context in the config they want to target!

Reply:
Totes agree, we should have the ability to specify (or at least pick up via env) a kube context.

Reply:
Great idea. Added!